src/codecs/pnm/decoder.rs

```diff
@@ -604,12 +647,17 @@ impl<R: Read> PnmDecoder<R> {
                 if bytes.len() != bytecount {
                     return Err(DecoderError::InputTooShort.into());
                 }
-
-                S::from_bytes(&bytes, width * components, buf)
-            }
-            SampleEncoding::Ascii => {
-                self.read_ascii::<S>(buf)
+                let row_size = match width.checked_mul(components.try_into().unwrap()) {
+                    Some(n) => n,
+                    None => return Err(ImageError::IoError(io::ErrorKind::UnexpectedEof.into())),
+                };
```

Comment:
Combining `unwrap` and `checked_mul` seems inconsistent. Note that not all platforms _guarantee_ that `usize` can be converted to `u32` and vice-versa. Rather the failure to conversion could also result in an error return.